tensorflow/c/kernels.cc

```diff
@@ -551,3 +554,257 @@ TF_Tensor* TF_AllocateTemp(TF_OpKernelContext* context, TF_DataType dtype,
+  using namespace tensorflow;
+  auto* context = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  if (var->copy_on_read_mode.load()) {
+    return Status::OK();
+  }
+  mutex_lock ml(*var->mu());
+  // Once copy-on-read mode is True the refcount is guaranteed to be 1. This can
+  // also happen if there are no concurrent reads of the variable and
+  // copy-on-read mode is false.
+  if (var->tensor()->RefCountIsOne()) {
+    var->copy_on_read_mode.store(true);
+    return Status::OK();
+  }
+  Tensor tmp;
+  if (variantType) {
+    AllocatorAttributes attr;
+    attr.set_on_host(true);
+    TF_RETURN_IF_ERROR(context->allocate_temp(var->tensor()->dtype(),
+                                          var->tensor()->shape(), &tmp, attr));
+
+    const auto elements_in = var->tensor()->flat<Variant>();
+    auto elements_out = tmp.flat<Variant>();
+    for (int64 i = 0; i < elements_in.size(); ++i) {
+      elements_out(i) = elements_in(i);
+    }
+  } else {
+    AllocatorAttributes attr;
+    attr.set_gpu_compatible(true);
+    attr.set_nic_compatible(true);
+    TF_RETURN_IF_ERROR(context->allocate_temp(var->tensor()->dtype(),
+                                          var->tensor()->shape(), &tmp, attr));
+    tensorflow::Status s;
+    TF_Tensor *tf_tmp = TF_TensorFromTensor(tmp, &s);
+    TF_Tensor *tf_tensor = TF_TensorFromTensor(*var->tensor(), &s);
+    copyFunc(ctx, tf_tensor, tf_tmp);
+  }
+  *var->tensor() = tmp;
+  var->copy_on_read_mode.store(true);
+  return Status::OK();
+}
+
+tensorflow::Status PrepareToUpdateVariable(TF_OpKernelContext* ctx, tensorflow::Tensor* tensor,
+                               bool copy_on_read_mode,
+                               bool variantType,
+                               void (*copyFunc)(TF_OpKernelContext * ctx, TF_Tensor *source, TF_Tensor *dest)) {
+
+  using namespace tensorflow;
+  auto* context = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  if (copy_on_read_mode || !tensor->RefCountIsOne()) {
+    // Tensor's buffer is in use by some read, so we need to copy before
+    // updating.
+    Tensor tmp;
+    if(variantType) {
+      AllocatorAttributes attr;
+      attr.set_on_host(true);
+      TF_RETURN_IF_ERROR(
+          context->allocate_temp(tensor->dtype(), tensor->shape(), &tmp, attr));
+
+      const auto elements_in = tensor->flat<Variant>();
+      auto elements_out = tmp.flat<Variant>();
+      for (int64 i = 0; i < elements_in.size(); ++i) {
+        elements_out(i) = elements_in(i);
+      }
+    } else {
+      AllocatorAttributes attr;
+      attr.set_gpu_compatible(true);
+      attr.set_nic_compatible(true);
+      TF_RETURN_IF_ERROR(
+          context->allocate_temp(tensor->dtype(), tensor->shape(), &tmp, attr));
+      tensorflow::Status s;
+      TF_Tensor *tf_tmp = TF_TensorFromTensor(tmp, &s);
+      TF_Tensor *tf_tensor = TF_TensorFromTensor(*tensor, &s);
+      copyFunc(ctx, tf_tensor, tf_tmp);
+    }
+    *tensor = tmp;
```

Comment:
The same here, tf_tmp tf_tensor should be deleted.

Reply:
Added.